workflow/controller/taskset.go

```diff
@@ -0,0 +1,204 @@
+package controller
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+	apiv1 "k8s.io/api/core/v1"
+	apierr "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/utils/pointer"
+
+	"github.com/argoproj/argo-workflows/v3/errors"
+	"github.com/argoproj/argo-workflows/v3/pkg/apis/workflow"
+	wfv1 "github.com/argoproj/argo-workflows/v3/pkg/apis/workflow/v1alpha1"
+	errorsutil "github.com/argoproj/argo-workflows/v3/util/errors"
+	"github.com/argoproj/argo-workflows/v3/util/retry"
+	waitutil "github.com/argoproj/argo-workflows/v3/util/wait"
+	"github.com/argoproj/argo-workflows/v3/workflow/common"
+)
+
+func (woc *wfOperationCtx) getAgentPodName() string {
+	return woc.wf.NodeID("agent") + "-agent"
+}
+
+func (woc *wfOperationCtx) executeTaskSet(ctx context.Context, nodeName string, templateScope string, tmpl *wfv1.Template, orgTmpl wfv1.TemplateReferenceHolder, opts *executeTemplateOpts) (*wfv1.NodeStatus, error) {
+	node := woc.wf.GetNodeByName(nodeName)
+	if node == nil {
+		node = woc.initializeExecutableNode(nodeName, wfv1.NodeTypeHTTP, templateScope, tmpl, orgTmpl, opts.boundaryID, wfv1.NodePending)
+	}
+	mainCtr := woc.newExecContainer(common.MainContainerName, tmpl)
+	mainCtr.Command = []string{"argoexec", "agent"}
+	// Append Agent in end of pod name to differentiate from normal podname
+	podName := woc.getAgentPodName()
+	_, err := woc.createAgentPod(ctx, podName, []apiv1.Container{*mainCtr}, tmpl)
+	if err != nil {
+		return woc.requeueIfTransientErr(err, node.Name)
+	}
+	if node.Phase == wfv1.NodePending {
```

Comment:
when would we create a non-pending task?

Reply:
Http template status will be updated to running and succeed based on Agent Pod status and Tastset status.

Reply:
But you are right. The non-pending node will not reach here